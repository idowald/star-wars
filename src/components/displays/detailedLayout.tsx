import * as React from "react";
import { MainProps } from "../mainDisplay";

export const DetailedLayout = (
  props: Omit<MainProps, "activeDisplay" | "isLoading">
) => {
  return <div>detailed {JSON.stringify(props)}</div>;
};
