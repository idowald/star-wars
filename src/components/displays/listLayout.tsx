import * as React from "react";
import { MainProps } from "../mainDisplay";

export const ListLayout = (
  props: Omit<MainProps, "activeDisplay" | "isLoading">
) => {
  return <div>List {JSON.stringify(props)}</div>;
};
