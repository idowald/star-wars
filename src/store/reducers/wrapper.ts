import { reducerWithInitialState } from "typescript-fsa-reducers";
import { Query } from "../types/query";
import {
  changeDisplay,
  closeModal,
  openModal,
} from "../actions/action";
import {
  getAvailableResources,
  getResources,
  getSingleResource
} from "../actions/resources";

export type ResourceName =
  | "films"
  | "people"
  | "planets"
  | "species"
  | "vehicles"
  | "starships";
export const resourcesNames: { [key in ResourceName]: string } = {
  films: "",
  people: "",
  planets: "",
  species: "",
  vehicles: "",
  starships: ""
};
export const displays: Display[] = ["grid", "list", "detailed"];
export type Display = "grid" | "list" | "detailed";
export interface Wrapper {
  display: Display;
  query: Query;
  availableResources: typeof resourcesNames;
  message: string | null;
  requestStatus: { [key: string]: boolean }; //a map of all loading requests by timestamp
  isModalOpen: boolean;
}

export const initialWrapper: Wrapper = {
  display: "grid",
  availableResources: resourcesNames,
  message: null,
  query: { search: null, id: null, pageNumber: null, resourceName: null },
  requestStatus: {},
  isModalOpen: false
};
export const wrapper = reducerWithInitialState(initialWrapper);
export const getAvailableResourcedReducer  = ({ requestStatus, ...state }:Wrapper, { timestamp }: {timestamp : string} ):Wrapper => {
  return {
    ...state,
    requestStatus: { ...requestStatus, [timestamp]: true }
  };
};
wrapper.case(
  getAvailableResources.started,
  getAvailableResourcedReducer
);
export const getAvailableResourcesFailed = ({ requestStatus, ...state }: Wrapper, { params: { timestamp }  }: {params : {timestamp : string}}) => {
  return {
    ...state,
    requestStatus: { ...requestStatus, [timestamp]: false }
  };
};
wrapper.case(
  getAvailableResources.failed,
    getAvailableResourcesFailed
);

wrapper.case(
  getAvailableResources.done,
  (
    { query, requestStatus, ...state },
    { result: { resourceNames }, params: { timestamp } }
  ) => {
    const firstAvailableResource = Object.keys(resourceNames).find(
      resource => !!resource
    ) as ResourceName;
    return {
      ...state,
      query: { ...query, resourceName: firstAvailableResource },
      availableResources: resourceNames,
      requestStatus: { ...requestStatus, [timestamp]: false }
    };
  }
);
wrapper.cases(
  [getResources.started, getSingleResource.started],
  ({ query, requestStatus, ...state }, { timestamp, ...newQuery }) => {
    return {
      ...state,
      query: { ...query, ...newQuery },
      requestStatus: { ...requestStatus, [timestamp]: true }
    };
  }
);
wrapper.cases(
  [getResources.failed, getSingleResource.failed],
  (
    { requestStatus, ...state },
    { error: { message }, params: { timestamp } }
  ) => {
    return {
      ...state,
      message,
      requestStatus: { ...requestStatus, [timestamp]: false }
    };
  }
);
wrapper.cases(
  [getResources.done, getSingleResource.done],
  ({ requestStatus, ...state }, { params: { timestamp } }) => {
    return {
      ...state,
      requestStatus: { ...requestStatus, [timestamp]: false }
    };
  }
);

wrapper.case(changeDisplay, (state: Wrapper, { display }) => ({
  ...state,
  display
}));
wrapper.case(openModal, (state: Wrapper) => ({ ...state, isModalOpen: true }));
wrapper.case(closeModal, (state: Wrapper) => ({
  ...state,
  isModalOpen: false
}));
