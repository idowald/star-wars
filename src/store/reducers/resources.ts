import { reducerWithInitialState } from "typescript-fsa-reducers";
import { Film } from "../types/film";
import { People } from "../types/people";
import { Planet } from "../types/planet";
import { Species } from "../types/species";
import { Starship } from "../types/starship";
import { Vehicle } from "../types/vehicle";
import { ResourceName } from "./wrapper";
import { getResources } from "../actions/resources";

export interface ResourceList<T> {
  name: ResourceName;
  resources: T[];
  next: string | null;
  previous: string | null;
  schema: { [property: string]: { type: string; format: string | null } } | {};
}
export type ResourceListState = ResourceList<
  Film | People | Planet | Species | Starship | Vehicle
>;

export const initialResources: ResourceListState = {
  name: "people",
  resources: [],
  next: null,
  previous: null,
  schema: {}
};
export const resources = reducerWithInitialState(initialResources);
resources.case(getResources.done, (state, { result }) => ({ ...result }));
