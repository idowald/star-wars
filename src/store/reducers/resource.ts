import { reducerWithInitialState } from "typescript-fsa-reducers";
import { Film } from "../types/film";
import { People } from "../types/people";
import { Planet } from "../types/planet";
import { Species } from "../types/species";
import { Starship } from "../types/starship";
import { Vehicle } from "../types/vehicle";
import { ResourceName } from "./wrapper";
import { getSingleResource } from "../actions/resources";
import { initialResources } from "./resources";

export interface Resource<
  T extends Film | People | Planet | Species | Starship | Vehicle
> {
  resourceName: ResourceName;
  resource: T | null;
  schema: typeof initialResources["schema"];
}
export type ResourceState = Resource<
  Film | People | Planet | Species | Starship | Vehicle
>;

export const initialResource: ResourceState = {
  resourceName: "people",
  resource: null,
  schema: {}
};
export const resource = reducerWithInitialState(initialResource);
// @ts-ignore
resource.case(getSingleResource.done, (state, { params, ...resource }) => {
  return { ...resource };
});
