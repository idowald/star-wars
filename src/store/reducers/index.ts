import { combineReducers } from "redux";
import { wrapper, initialWrapper } from "./wrapper";
import { initialResources, resources } from "./resources";
import { initialResource, resource } from "./resource";

export interface State {
  wrapper: typeof initialWrapper;
  resources: typeof initialResources;
  resource: typeof initialResource;
}

export const reducer = combineReducers<State>({
  wrapper,
  resources,
  resource
});
