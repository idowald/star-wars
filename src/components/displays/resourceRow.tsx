import { Icon, TableCell } from "@material-ui/core";
import { typesRelations } from "../../utils/schema";
import * as React from "react";
import { ResourceState } from "../../store/reducers/resource";
import { MainProps } from "../mainDisplay";
import { typesIcons } from "../../utils/icons";

interface Props {
  resource: ResourceState["resource"];
  openModal: MainProps["openModal"];
}
export const ResourceRow = ({ resource, openModal }: Props) => {
  if (!resource) {
    return <React.Fragment></React.Fragment>;
  }
  const properties = Object.keys(resource).filter(
    key => key !== "created" && key !== "edited" && key !== "url"
  );
  return (
    <React.Fragment>
      <TableCell>{new Date(resource.created).toDateString()}</TableCell>
      <TableCell>{new Date(resource.edited).toDateString()}</TableCell>
      {properties.map((key: string) => {
        // @ts-ignore
        const value = resource[key];
        if (typesRelations.hasOwnProperty(key)) {
          return (
            <TableCell key={key}>
              {Array.isArray(value) ? (
                value.map(relation => (
                  <span key={relation} onClick={() => openModal(relation)}>
                    <Icon>{typesIcons[key]}</Icon>
                  </span>
                ))
              ) : (
                <span onClick={() => openModal(value)}>
                  <Icon>{typesIcons[key]}</Icon>
                </span>
              )}
            </TableCell>
          );
        }
        // Note- I just wanted to show a way that we can render numbers from string, and do manipulation on them if needed
        const number = Number(value);
        if (isNaN(number)) {
          return <TableCell key={key}>{value}</TableCell>;
        } else {
          return <TableCell key={key}>{number}</TableCell>;
        }
      })}
    </React.Fragment>
  );
};
