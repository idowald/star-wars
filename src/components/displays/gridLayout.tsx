import * as React from "react";
import { MainProps } from "../mainDisplay";
import {
  TableContainer,
  Paper,
  Table,
  TableHead,
  TableCell,
  TableBody,
  TableRow,
} from "@material-ui/core";
import { ResourceRow } from "./resourceRow";

export const GridLayout = (
  props: Omit<MainProps, "activeDisplay" | "isLoading">
) => {
  let properties: string[] = [];
  if (props.resources && props.resources.resources.length) {
    properties = Object.keys(props.resources.resources[0]).filter(
      key => key !== "created" && key !== "edited" && key !== "url"
    );
  }
  if (props.resource && props.resource.resource) {
    properties = Object.keys(props.resource.resource).filter(
      key => key !== "created" && key !== "edited" && key !== "url"
    );
  }

  return (
    <TableContainer component={Paper}>
      <Table aria-label="table">
        <TableHead>
          <TableRow>
            <TableCell>created</TableCell>
            <TableCell>edited</TableCell>
            {properties.map(property => (
              <TableCell key={property}>{property}</TableCell>
            ))}
          </TableRow>
        </TableHead>
        <TableBody>
          {props.resource ? (
            <TableRow>
              <ResourceRow
                resource={props.resource.resource}
                openModal={props.openModal}
              />
            </TableRow>
          ) : (
            props.resources &&
            props.resources.resources.map(resource => (
              <TableRow key={resource.url}>
                <ResourceRow resource={resource} openModal={props.openModal} />
              </TableRow>
            ))
          )}
        </TableBody>
      </Table>
    </TableContainer>
  );
};
