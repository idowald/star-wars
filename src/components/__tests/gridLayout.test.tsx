import React from "react";
import renderer from "react-test-renderer";
import {GridLayout} from "../displays/gridLayout";
import {ResourceState} from "../../store/reducers/resource";
import {ResourceName} from "../../store/reducers/wrapper";


const resources = {
  name: 'people' as ResourceName,
  resources: [
    {
      name: 'Luke Skywalker',
      height: '172',
      mass: '77',
      hair_color: 'blond',
      skin_color: 'fair',
      eye_color: 'blue',
      birth_year: '19BBY',
      gender: 'male',
      homeworld: 'http://swapi.dev/api/planets/1/',
      films: [
        'http://swapi.dev/api/films/1/',
        'http://swapi.dev/api/films/2/',
        'http://swapi.dev/api/films/3/',
        'http://swapi.dev/api/films/6/'
      ],
      species: [],
      vehicles: [
        'http://swapi.dev/api/vehicles/14/',
        'http://swapi.dev/api/vehicles/30/'
      ],
      starships: [
        'http://swapi.dev/api/starships/12/',
        'http://swapi.dev/api/starships/22/'
      ],
      created: '2014-12-09T13:50:51.644000Z',
      edited: '2014-12-20T21:17:56.891000Z',
      url: 'http://swapi.dev/api/people/1/'
    },
    {
      name: 'C-3PO',
      height: '167',
      mass: '75',
      hair_color: 'n/a',
      skin_color: 'gold',
      eye_color: 'yellow',
      birth_year: '112BBY',
      gender: 'n/a',
      homeworld: 'http://swapi.dev/api/planets/1/',
      films: [
        'http://swapi.dev/api/films/1/',
        'http://swapi.dev/api/films/2/',
        'http://swapi.dev/api/films/3/',
        'http://swapi.dev/api/films/4/',
        'http://swapi.dev/api/films/5/',
        'http://swapi.dev/api/films/6/'
      ],
      species: [
        'http://swapi.dev/api/species/2/'
      ],
      vehicles: [],
      starships: [],
      created: '2014-12-10T15:10:51.357000Z',
      edited: '2014-12-20T21:17:50.309000Z',
      url: 'http://swapi.dev/api/people/2/'
    },
    {
      name: 'R2-D2',
      height: '96',
      mass: '32',
      hair_color: 'n/a',
      skin_color: 'white, blue',
      eye_color: 'red',
      birth_year: '33BBY',
      gender: 'n/a',
      homeworld: 'http://swapi.dev/api/planets/8/',
      films: [
        'http://swapi.dev/api/films/1/',
        'http://swapi.dev/api/films/2/',
        'http://swapi.dev/api/films/3/',
        'http://swapi.dev/api/films/4/',
        'http://swapi.dev/api/films/5/',
        'http://swapi.dev/api/films/6/'
      ],
      species: [
        'http://swapi.dev/api/species/2/'
      ],
      vehicles: [],
      starships: [],
      created: '2014-12-10T15:11:50.376000Z',
      edited: '2014-12-20T21:17:50.311000Z',
      url: 'http://swapi.dev/api/people/3/'
    },

  ],
  next: 'http://swapi.dev/api/people/?page=2',
  previous: null,
  schema: {
    starships: {
      type: 'array',
      description: 'An array of starship resources that this person has piloted'
    },
    edited: {
      type: 'string',

    },
    name: {
      type: 'string',

    },
    created: {
      type: 'string',

    },
    url: {
      type: 'string',

    },
    gender: {
      type: 'string',

    },
    vehicles: {
      type: 'array',
    },
    skin_color: {
      type: 'string',
    },
    hair_color: {
      type: 'string',
    },
    height: {
      type: 'string',
    },
    eye_color: {
      type: 'string',
    },
    mass: {
      type: 'string',
    },
    films: {
      type: 'array',
    },
    species: {
      type: 'array',
    },
    homeworld: {
      type: 'string',
    },
    birth_year: {
      type: 'string',
    }
  }
};
const resource :ResourceState = {resource:  null, resourceName: "people", schema :{}}
test("test grid display for snapshot", () => {
  const rendered = renderer.create(
    <GridLayout
      openModal={()=>{}}
      resource={resource}
      resources={resources}
    />
  );
  expect(rendered).toMatchSnapshot();
});
