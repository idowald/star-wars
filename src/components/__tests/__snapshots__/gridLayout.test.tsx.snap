// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test grid display for snapshot 1`] = `
<div
  className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <table
    aria-label="table"
    className="MuiTable-root"
    role={null}
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          created
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          edited
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          name
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          height
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          mass
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          hair_color
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          skin_color
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          eye_color
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          birth_year
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          gender
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          homeworld
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          films
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          species
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          vehicles
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          starships
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root"
        role={null}
      />
    </tbody>
  </table>
</div>
`;
